e2e/pollers.go

```diff
@@ -230,3 +236,49 @@ func waitUntilPodDeleted(ctx context.Context, kube *kubeclient, podName string)
 		return err == nil, err
 	})
 }
+
+type Poller[T any] interface {
+	PollUntilDone(ctx context.Context, options *runtime.PollUntilDoneOptions) (T, error)
+}
+
+func pollVMSSOperation[T any](ctx context.Context, vmssName string, pollerOpts *runtime.PollUntilDoneOptions, vmssOperation func() (Poller[T], error)) (*T, error) {
+	var requestError azure.RequestError
+	var err error
+	var poller Poller[T]
+	for i := 0; i < vmssRetryAttempts; i++ {
+		poller, err = vmssOperation()
+		if err != nil {
+			if errors.Is(err, context.DeadlineExceeded) {
+				return nil, fmt.Errorf("unable to complete VMSS operation in allotted time of %s: %w", createVMSSPollingTimeout.String(), err)
+			}
+			if errors.As(err, &requestError) && requestError.ServiceError != nil {
+				log.Printf("VMSS operation failed on attempt %d due to %v, retrying...", i+1, requestError)
+				time.Sleep(vmssRetrySleep)
+				continue
+			}
+			return nil, err
+		}
+		break
+	}
+	if err != nil {
+		return nil, fmt.Errorf("error on VMSS operation, all retries failed %q: %v", vmssName, err)
+	}
+
+	var vmssResp T
+	for i := 0; i < vmssRetryAttempts; i++ {
+		vmssResp, err = poller.PollUntilDone(ctx, pollerOpts)
```

Comment:
I don't think retrying the polling here makes sense right? I believe the underlying poller returned by the ARM API will poll correctly until a terminal state has been reached - we need to check that terminal state and retry the actual operation itself if that terminal state isn't a success

Reply:
now that I think about it I don't believe we actually need to retry the actual invocation of the op to get the poller object - the ARM client retry config should take care of that. We really just care about retrying in the case where the terminal state reported by the poller obj is non-success. So we'd wanna do something like:


```suggestion
for retry in retries {
  poller, err := op()
  if err != nil {
    fail
  }

  resp, err := poller.PollUntilDone()
  if err != nil {
    if errors.As(...) && requestError.ServiceError != nil {
        maybe check the code
        sleep()
        continue
    }
    fail
  }
  
  return success
}

```